### DISCUSSION TITLE
models/yolo-nas/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9526] [Docs]
# models/yolo-nas/

Explore detailed documentation of YOLO-NAS, a superior object detection model. Learn about its features, pre-trained models, usage with Ultralytics Python API, and more.

https://docs.ultralytics.com/models/yolo-nas/

**Comment by Pragyy-G** [comment]
How to train out this Yolo-nas model? like fine tune it on custom dataset?

**Reply by glenn-jocher** [reply]
@Pragyy-G hey there! 🙌 Training YOLO-NAS on your custom dataset for fine-tuning involves a few steps since the provided models are primarily aimed at inference and validation — as noted on our [YOLO-NAS documentation page](https://docs.ultralytics.com/models/yolo-nas), direct support for training is not currently highlighted.

However, you generally have a couple of options:

1. **Utilize Transfer Learning:** You can leverage the pre-trained weights as a starting point and train on your custom dataset. This involves freezing the initial layers and retraining the later layers on your dataset, though the specific capability should be verified within the Ultralytics framework or custom implemented.

2. **Custom Training Routine:** Incorporate YOLO-NAS architecture within a custom training routine where you can specify your dataset, optimization parameters, and training loop. This is a bit advanced and requires a good understanding of PyTorch and the model architecture.

Since YOLO-NAS's training guidance is not explicit in the docs for custom datasets, here's a general approach you might consider, aligning with typical Ultralytics models:

```python
from ultralytics import YOLO, Dataset

# Load your custom dataset
your_dataset = Dataset('path/to/your/dataset')

# Initialize the model, ensuring to adjust for your specific use case
model = YOLO('yolo_nas_s.pt')  # Load YOLO-NAS with small variant for example

# Here you might need to adjust model for training if direct training is supported 
# This step is conceptual and might vary in implementation

# Define your training parameters (hypothetical example)
train_params = {
    'epochs': 100,
    'batch_size': 16,
    'lr': 0.001
}

# Train the model on your dataset
model.train(your_dataset, **train_params)
```

For fine-tuning YOLO models on custom datasets, it's crucial to check out the [Ultralytics documentation](https://docs.ultralytics.com/) and GitHub repo for the latest info and examples, as direct support for training specific NAS models may evolve.

If training YOLO-NAS directly is not feasible due to current tooling, exploring Ultralytics's other models (like YOLOv5 or YOLOv8) which have extensive support for custom training might be beneficial. They can serve as a robust foundation and offer a path to employ NAS principles in part.

Hope this helps a bit! Feel free to reach out for more detailed guidance based on your project's specifics. Happy coding! 😊

**Reply by Tottowich** [reply]
@Pragyy-G
Any results on this?

**Reply by ashish-reddy26** [reply]
Hi Glenn, 

I tried training with the provided information, but I am getting the error that says 'Dataset' cannot be imported from ultralytics.
Is Ultralytics not providing the resources to Dataset?

Can you clarify this and also please provide your inputs for training yolo-nas for object detection tasks (I am using yolov8 & yolov9, but I also want to try with yolo-nas as well).

Thank you,
Ashish

**Reply by glenn-jocher** [reply]
Hi Ashish,

It seems there might be a misunderstanding regarding the import statement. Currently, `Dataset` isn't a directly importable class from the `ultralytics` package. For custom training datasets, you typically need to structure your data according to the expected format and use a custom YAML file to define paths and settings.

For YOLO-NAS, as mentioned in the docs, it primarily supports inference and validation and does not directly support training out-of-the-box. If you're keen on training YOLO-NAS or any versions not originally designed for training, you might consider adapting your training script to handle NAS-specific configurations—or use models like YOLOv8 or YOLOv9 which are fully equipped for training on custom datasets.

Here’s a general outline on how you might set up a validation dataset, which might help in adjusting your approach:

```python
# Assuming you have structured your dataset properly
# Your data directory should include a 'data.yaml' file that defines the train/val path, classes, etc.

from ultralytics import YOLO

# Load the model
model = YOLO('path/to/yolo-nas-model.pt')

# Validate the model
results = model.val(data='path/to/your/data.yaml')
```

Please ensure your dataset follows the structure expected by the Ultralytics framework and that the YAML file is correctly pointing to your train and validation images and labels.

Hope this helps, and thank you for trying out YOLO-NAS alongside YOLOv8 and YOLOv9!

**Comment by riobastian09** [comment]
Hi. i want to know is there any pre-configured of YOLO-NAS model? i want to use this to my experiment using the custom architecture YOLO-NAS model. Thank you!

**Reply by glenn-jocher** [reply]
@riobastian09 thank you for your question! Yes, pre-configured YOLO-NAS models, such as YOLO-NAS-s, YOLO-NAS-m, and YOLO-NAS-l, are available and optimized for inference, validation, and export. You can access these pre-trained models and easily integrate them into your experiments using the Ultralytics Python package. Unfortunately, custom architecture modifications or training directly with YOLO-NAS are not supported. For more, refer to the YOLO-NAS documentation: https://docs.ultralytics.com/models/yolo-nas/
